plugin.video.putio/main.py

```diff
@@ -0,0 +1,181 @@
+
+# put.io kodi addon
+# Copyright (C) 2009  Alper Kanat <alper@put.io>
+# Copyright (C) 2016  Put.io Developers <devs@put.io>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import sys
+import urlparse
+import requests
+
+import xbmc
+import xbmcgui
+import xbmcplugin
+
+from resources.lib.helper import __lang__
+from resources.lib.helper import __settings__
+from resources.lib.helper import PutioApiHandler
+from resources.lib.helper import PutioAuthFailureException
+
+__url__ = sys.argv[0]  # base URL ('plugin://plugin.video.putio/')
+__handle__ = int(sys.argv[1])  # process handle, as a numeric string
+__args__ = urlparse.parse_qs(sys.argv[2].lstrip('?'))  # query string, ('?action=list&item=3')
+
+PUTIO_KODI_ENDPOINT = 'https://put.io/xbmc'
+RESOURCE_PATH = os.path.join(__settings__.getAddonInfo('path'), 'resources', 'media')
+
+
+def build_url(action, item):
+    return '{0}?action={1}&item={2}'.format(__url__, action, item)
+
+
+def get_resource_path(filename):
+    """Returns special path of the given filename."""
+    if not filename:
+        return
+    return os.path.join(RESOURCE_PATH, filename)
+
+
+def populate_dir(files):
+    """Fills a directory listing with put.io files."""
+    list_items = []
+    for item in files:
+        thumbnail = item.screenshot or get_resource_path('mid-folder.png')
+        li = xbmcgui.ListItem(label=item.name,
+                              label2=item.name,
+                              iconImage=thumbnail,
+                              thumbnailImage=thumbnail)
+
+        # url when a delete action is triggered
+        delete_ctx_url = build_url(action='delete', item=item.id)
+
+        if item.is_folder:
+            url = build_url(action='list', item=item.id)
+        else:  # video or audio, no other types are available here
+            url = build_url(action='play', item=item.id)
+
+            # resumetime and totaltime are needed for Kodi to decide the file as watched or not.
+            # FIXME: get total-time of the file and set to 'totaltime'
+            li.setProperty(key='resumetime', value=str(item.start_from))
+            li.setProperty(key='totaltime', value=str(20 * 60))
+
+            # http://kodi.wiki/view/InfoLabels
+            type_ = 'video' if item.is_video else 'audio'
+            li.setInfo(type=type_, infoLabels={'size': item.size, 'title': item.name})
```

Comment:
plase also set a valid mediatype (see http://mirrors.kodi.tv/docs/python-docs/16.x-jarvis/xbmcgui.html#ListItem-setInfo ) if possible in order to fix issues with newer kodi versions.